sdk/python/feast/infra/offline_stores/bigquery.py

```diff
@@ -310,6 +325,25 @@ def _execute_query(
         block_until_done(client=self.client, bq_job=bq_job, timeout=timeout)
         return bq_job
 
+    def persist(self, storage: SavedDatasetStorage) -> "BigQueryRetrievalJob":
```

Comment:
Do we want this to be a publically facing method? Seems confusing to a user why there'd be

1. this persist method, which happens lazily. 
2. `to_bigquery`, which exists right next to this method and also saves.

I think the return type can be confusing too since .persist has indeed executed a query & persisted, but we return a `RetrievalJob`. I might as a user expect to need to call `RetrievalJob#to_bigquery` again, which actually will persist to a different location.

Two options I think could make sense:
- Have this be an internal only method by name so people don't try to call it
- Expose a separate way to instantiate a RetrievalJob (like RetrievalJob.from_saved_dataset which executes this wrapped query) so you can return it in `get_historical_features`

Reply:
`persist` doesn't happen lazily. It's generic equivalent of methods `to_bigquery` or `to_redshift` which exist only in specific implementation classes. `predict` is part of abstract API (`RetrievalJob`) and aimed to unify these methods. Under the hood, it's calling `to_bigquery` or `to_redshift`. 

The feature that is absent in previous methods is that `persist` returns RetrievalJob, which supposedly retrieves from a persisted table. 

Why it's public? As we discussed on call with Shopify - it could be useful if the user doesn't want `get_historical_features` to run query eagerly. In this case, user would not pass `save_as` parameter but call `persist` themselves:
```
job = fs.get_historical_feature(...)
persisted = job.persist(storage=SavedDatasetBigQueryStorage(...))
persisted.to_df()
```

Regarding cyclical calls, I think that we could deprecate specific methods like `.to_bigquery()` and then it should be less confusing.

> Expose a separate way to instantiate a RetrievalJob (like RetrievalJob.from_saved_dataset which executes this wrapped query) so you can return it in get_historical_features

This option makes sense to me. Please let me know what you think is better @adchia

Reply:
UPD. `persist` signature updated. Now it doesn't return `RetrievalJob` to avoid duplication of responsibilities. Read of saved dataset moved to Provider.read_saved_dataset, which uses OfflineSource.pull_all_from_table_or_query.

So now `.persist()` is just a universal replacement for `.to_bigquery()` and `.to_redshift()`.